src/function/execute.rs

```diff
@@ -38,53 +36,114 @@ where
+                let last_provisional_value = if let Some(last_provisional) = opt_last_provisional {
+                    // We have a last provisional value from our previous time around the loop.
+                    &last_provisional
+                        .value
+                        .as_ref()
+                        .expect("provisional value evicted by LRU?")
+                } else {
+                    // This is our first time around the loop; a provisional value must have been
+                    // inserted into the memo table when the cycle was hit, so let's pull our
+                    // initial provisional value from there.
+                    opt_owned_last_provisional = self.get_memo_from_table_for(zalsa, id);
+                    &opt_owned_last_provisional
+                        .as_deref()
+                        .expect(
+                            "{database_key_index:#?} is a cycle head, \
+                            but no provisional memo found",
+                        )
+                        .value
+                        .as_ref()
+                        .expect("provisional value evicted by LRU?")
+                };
                 tracing::debug!(
-                    "{database_key_index:?}: caught cycle {cycle:?}, have strategy {:?}",
-                    C::CYCLE_STRATEGY
+                    "{database_key_index:?}: execute: \
+                            I am a cycle head, comparing last provisional value \
+                            {last_provisional_value:#?} with new value {new_value:#?}"
                 );
-                match C::CYCLE_STRATEGY {
-                    crate::cycle::CycleRecoveryStrategy::Panic => cycle.throw(),
-                    crate::cycle::CycleRecoveryStrategy::Fallback => {
-                        if let Some(c) = active_query.take_cycle() {
-                            assert!(c.is(&cycle));
-                            C::recover_from_cycle(db, &cycle, C::id_to_input(db, id))
-                        } else {
-                            // we are not a participant in this cycle
-                            debug_assert!(!cycle
-                                .participant_keys()
-                                .any(|k| k == database_key_index));
-                            cycle.throw()
+                // If the new result is equal to the last provisional result, the cycle has
+                // converged and we are done.
+                if !C::values_equal(&new_value, last_provisional_value) {
+                    // We are in a cycle that hasn't converged; ask the user's
+                    // cycle-recovery function what to do:
+                    match C::recover_from_cycle(db, &new_value, iteration_count) {
+                        crate::CycleRecoveryAction::Iterate => {
+                            tracing::debug!("{database_key_index:?}: execute: iterate again");
+                            iteration_count += 1;
```

Comment:
Let's panic with a meaningful message if the algorithm doesn't converage in `u32::MAX` steps


```suggestion
                            iteration_count = iteration_count.checked_add(1).expect("Cycle recovery to converge after u32::MAX iterations");
```